### ISSUE TITLE
 Regression Testing Setup

### ISSUE THREAD

**Issue by MarcelHB** [gemrb/gemrb#1885]
I don't want to do TDD or refactor anything for code coverage but I think there is space to think about some automated testing and CI.

There are already smaller units and ones that are already properly isolated that can easily get a bunch of tests to make sure that the next wave of refactoring will not leave us blind towards a bunch of regressions. Worst case to me is corrupting savegames.

I'd like to discuss anything related to how and what, testing libs/frameworks, mildly convenient CMake integration.

**Comment by bradallred** [comment]
cmake has [CTest](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Testing%20With%20CMake%20and%20CTest.html)

**Comment by lynxlynxlynx** [comment]
Some of the low-level stuff we can definitely cover and it can help with making our code reusable (people struggled before). 

The first things that come to mind are VFS, FixedSizedString and other string stuff (Explode was a nightmare to get right), maybe some importers, and the minimal test we use on CI could be made into a CTest test, so it's easier to run.

**Comment by MarcelHB** [comment]
Yeah, CTest as driver makes sense for us. Preferences regarding invidual tests? I'm mildly familiar with GoogleTest but maybe there is also something more convincing.

> The first things that come to mind are VFS, FixedSizedString and other string stuff (Explode was a nightmare to get right), maybe some importers

That's what I'm also thinking about for the first cases.

**Comment by lynxlynxlynx** [comment]
I haven't used any in practice. A quick search also suggests a fast https://github.com/doctest/doctest (inlining support may make it easier to deploy) and https://github.com/catchorg/Catch2, both of which also come in a single header-only version.

**Comment by MarcelHB** [comment]
Hm, I'm playing around with all three of them since they are somewhat different (and GT comes with mocking facilities) and let's then have a look what looks scalable for us.

**Comment by bradallred** [comment]
Thanks for doing the research @MarcelHB. Tests will be very helpful, even if limited in scope.

**Comment by MarcelHB** [comment]
I've done a little setup in all three libs with CMake Test integration with cases of fresh memory. I even found the first bug this way (see VFS.cpp and Win32 test case) :slightly_smiling_face: 

What is nice: all three libs support putting a test section anywhere and one runner will include them all. So we only need one test in terms of CMake, not one for each unit. That also helps compilation speed.

This requires putting the cpp file explicitly to the test's sources. For the core, they are then recompiled again with the test section and linked against `gemrb_core` (this requires `WITH_TESTS` for now so that `gemrb_core` remains free of test stuff). There is maybe some precaution advised (see copying the DLL on Windows) when doing dynamic linking.

You _could_ also refrain from linking against `gemrb_core` and compile a test unit minimally with required units and ext. libs (iconv, WinAPI, ...) only but this collides when looking at our code base, i. e. `VFS` -> `Logger` -> `TextArea` -> `View` -> `...`. So that's infeasible but maybe an indication of overly big coupling to solve at the long term.

I haven't looked at how to work with the plugins as well yet.

So doctest probably suffices for the moment but I have no idea how good it is when trying more sophisticated tests. Catch2 would also do, I guess, and I don't exactly see how this is better or worse for us than doctest, expect maybe having a wider range of users.

GoogleTest lacks a little convenience but really makes sense to me if the code base follows some patterns that allow mocking of injected dependencies and check whether their interactions a proper. But otherwise, we are not a business application and it's big enough to require some more steps for setup in CMake.

**Comment by lynxlynxlynx** [comment]
Core.cpp and some of the other geometry stuff are also probably fitting targets to test.
I'm leaning towards gtest, thanks for the detailed exploration!

**Comment by bradallred** [comment]
gtest is what I use professionally, so its easy for me to get on board with that too.

**Comment by MarcelHB** [comment]
Then let's say Google Test.

I'll try to set up something with a focus on distinct test sources and maybe also some resources.

We can have a look at this then in a new PR.
